shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java

```diff
@@ -0,0 +1,217 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.embedding.engine;
+
+import android.content.res.AssetManager;
+import android.graphics.Bitmap;
+import android.graphics.SurfaceTexture;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.util.Log;
+import android.view.Surface;
+
+import java.nio.ByteBuffer;
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArraySet;
+
+import io.flutter.embedding.engine.dart.PlatformMessageHandler;
+import io.flutter.embedding.engine.FlutterEngine.EngineLifecycleListener;
+import io.flutter.embedding.engine.renderer.FlutterRenderer;
+import io.flutter.embedding.engine.renderer.OnFirstFrameRenderedListener;
+
+/**
+ * WARNING: THIS CLASS IS EXPERIMENTAL. DO NOT SHIP A DEPENDENCY ON THIS CODE.
+ * IF YOU USE IT, WE WILL BREAK YOU.
+ */
+public class FlutterJNI {
+  private static final String TAG = "FlutterJNI";
+
+  private FlutterRenderer.RenderSurface renderSurface;
+  private PlatformMessageHandler platformMessageHandler;
+  private final Set<EngineLifecycleListener> engineLifecycleListeners = new CopyOnWriteArraySet<>();
```

Comment:
Are we expected to modify this from different threads?
If so would be nice to note in e.g addEngineLifecycleListener that this is thread safe.